mixer/test/client/istio_authn/authn_report_test.go

```diff
@@ -0,0 +1,323 @@
+  "mesh2.ip": "[0 0 0 0 0 0 0 0 0 0 255 255 204 152 189 116]",
+  "mesh3.ip": "[0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 8]",
+  "request.host": "*",
+  "request.path": "/echo",
+  "request.time": "*",
+  "request.useragent": "Go-http-client/1.1",
+  "request.method": "GET",
+  "request.scheme": "http",
+  "source.uid": "POD11",
+  "source.namespace": "XYZ11",
+  "source.ip": "[127 0 0 1]",
+  "source.port": "*",
+  "destination.ip": "[127 0 0 1]",
+  "destination.port": "*",
+  "target.name": "target-name",
+  "target.user": "target-user",
+  "target.uid": "POD222",
+  "target.namespace": "XYZ222",
+  "connection.mtls": false,
+  "check.cache_hit": false,
+  "quota.cache_hit": false,
+  "request.headers": {
+     ":method": "GET",
+     ":path": "/echo",
+     ":authority": "*",
+     "x-forwarded-proto": "http",
+     "x-istio-attributes": "-",
+     "x-request-id": "*"
+  },
+  "request.size": 0,
+  "request.total_size": 517,
+  "response.total_size": 99,
+  "response.time": "*",
+  "response.size": 0,
+  "response.duration": "*",
+  "response.code": 200,
+  "response.headers": {
+     "date": "*",
+     "content-length": "0",
+     ":status": "200",
+     "server": "envoy"
+  },
+  "request.auth.audiences": "aud1",
+  "request.auth.claims": {
+     "iss": "issuer@foo.com", 
+     "sub": "sub@foo.com",
+     "aud": "aud1",
+     "some-other-string-claims": "some-claims-kept"
+  }
+}
+`
+
+// TODO: convert to v2, real clients use bootstrap v2 and all configs are switching !!!
+// The envoy config template with Istio authn filter
+const envoyConfTempl = `
+{
+  "listeners": [
+    {
+      "address": "tcp://0.0.0.0:{{.ServerPort}}",
+      "bind_to_port": true,
+      "filters": [
+        {
+          "type": "read",
+          "name": "http_connection_manager",
+          "config": {
+            "codec_type": "auto",
+            "stat_prefix": "ingress_http",
+            "route_config": {
+              "virtual_hosts": [
+                {
+                  "name": "backend",
+                  "domains": ["*"],
+                  "routes": [
+                    {
+                      "timeout_ms": 0,
+                      "prefix": "/",
+                      "cluster": "service1",
+                      "opaque_config": {
+{{.MixerRouteFlags}}
+                      }
+                    }
+                  ]
+                }
+              ]
+            },
+            "access_log": [
+              {
+                "path": "{{.AccessLog}}"
+              }
+            ],
+            "filters": [
+{{.FaultFilter}}
+              {
+                "type": "decoder",
+                "name": "istio_authn",
+                "config": {
+                  "policy": {
+                    "origins": [
+                      {
+                        "jwt": {
+                          "issuer": "issuer@foo.com",
+                          "jwks_uri": "http://localhost:8081/"
+                        }
+                      }
```

Comment:
Need another test for with binding to origin. Also, can we have another for peer / mTLS case?

Reply:
Added the tests with binding to origin. For peer/mTLS tests, I prefer to add them in separate PRs.

Reply:
Added peer, TLS, and mTLS tests. The following is the tests added in this PR: 
- Test when requiring JWT for peer and binding to peer, the authn attributes in the actual check and report calls match those in the expected check call
- Test when requiring JWT for peer and binding to origin, but no method specified in origin policy, the request will be rejected by Istio authn filter.
- Test when when requiring JWT for origin and binding to origin, the authn attributes in the actual check and report calls match those in the expected check call.
- Test when requiring JWT for origin and no binding to origin, the authn attributes in the actual check and report calls match those in the expected check call.
- Test when the HTTP request is rejected by the Istio authn filter for peer JWT authentication, the response code and the response message is as expected.
- Test when the HTTP request is rejected by the Istio authn filter for origin JWT authentication, the response code and the response message is as expected.
- Test when the Istio authn filter requires mTLS for peer connection, the non mTLS connection is rejected and the response code and the response message are as expected.
- Test when the Istio authn filter requires TLS for peer connection, the non TLS connection is rejected and the response code and the response message are as expected.